Content.Server/Chat/V2/Messages.cs

```diff
@@ -92,3 +92,50 @@ public sealed class WhisperCreatedEvent(EntityUid speaker, string message, float
     public float MaxRange = maxRange;
 }
 
+/// <summary>
+/// Notifies that a chat message needs validating.
+/// </summary>
+/// <param name="attemptEvent">The chat message to validate</param>
+public sealed class ChatValidationEvent<T>(T attemptEvent) where T : ChatAttemptEvent
+{
+    public readonly T Event = attemptEvent;
+    public string Reason = "";
+
+    public bool IsCancelled { get; private set; }
+
+    public void Cancel(string reason)
+    {
+        if (IsCancelled)
+        {
+            return;
+        }
+
+        IsCancelled = true;
+        Reason = reason;
+    }
+}
+
+/// <summary>
+/// Notifies that a chat message needs sanitizing. If, after this message is processed, IsCancelled is true, the message
+/// should be discarded with a failure response. Otherwise, if ChatMessageSanitized is non-null, ChatMessageSanitized
+/// should be used instead of the non-sanitized message.
+/// </summary>
+/// <param name="attemptEvent">The chat message to sanitize.</param>
+[ByRefEvent]
+public sealed class ChatSanitizationEvent<T>(T attemptEvent) where T : ChatAttemptEvent
```

Comment:
Local events should be readonly structs but only if they ARE readonly. if its not readonly (i see mutation) - then remove ByRefEvent i guess?